File: app/src/main/java/com/meteomontana/android/ui/components/GradeFilterBar.kt
package com.meteomontana.android.ui.components

import androidx.compose.animation.animateContentSize
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.KeyboardArrowDown
import androidx.compose.material.icons.outlined.KeyboardArrowRight
import androidx.compose.material.icons.outlined.KeyboardArrowUp
import androidx.compose.material.icons.outlined.Tune
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.meteomontana.android.domain.util.GradeFilterResult
import com.meteomontana.android.domain.util.GradeMatch
import com.meteomontana.android.ui.theme.EyebrowTextStyle
import com.meteomontana.android.ui.theme.Ink
import com.meteomontana.android.ui.theme.Ink2
import com.meteomontana.android.ui.theme.Ink3
import com.meteomontana.android.ui.theme.Paper
import com.meteomontana.android.ui.theme.Rule
import com.meteomontana.android.ui.theme.Terra
import com.meteomontana.android.ui.theme.colorForGrade

// Filtro LOCAL por grado dentro de una escuela — ver BLOCK_SEARCH_DESIGN.md §7.
// Espejo de GradeFilterBar.swift (iOS): chips multi-selección con la paleta
// de grados de la app (mismo patrón que el diario), resultados agrupados por
// grado y colapsables, tocar una vía abre su piedra. La lógica de qué
// piedra/vía cae en la selección vive en shared (GradeFilter.kt).

@Composable
fun GradeFilterBar(
    selectedGrades: Set<String>,
    onSelectedGradesChange: (Set<String>) -> Unit,
    availableGrades: List<String>,
    result: GradeFilterResult,
    onSelectLine: (GradeMatch) -> Unit
) {
    var expanded by remember { mutableStateOf(false) }
    var openGroups by remember { mutableStateOf(setOf<String>()) }
    val isActive = selectedGrades.isNotEmpty()

    Column(Modifier.fillMaxWidth().animateContentSize()) {
        Row(
            Modifier
                .fillMaxWidth()
                .clickable { expanded = !expanded }
                .padding(horizontal = 16.dp, vertical = 8.dp),
            verticalAlignment = androidx.compose.ui.Alignment.CenterVertically
        ) {
            Icon(Icons.Outlined.Tune, contentDescription = null,
                tint = if (isActive) Terra else Ink3, modifier = Modifier.width(16.dp))
            Spacer(Modifier.width(6.dp))
            Text("FILTRAR POR GRADO", style = EyebrowTextStyle.copy(fontSize = 10.sp),
                color = if (isActive) Terra else Ink3)
            Spacer(Modifier.weight(1f))
            if (isActive) {
                Text("QUITAR TODO", style = EyebrowTextStyle.copy(fontSize = 9.sp), color = Ink3,
                    modifier = Modifier.clickable { onSelectedGradesChange(emptySet()) })
                Spacer(Modifier.width(10.dp))
            }
            Icon(
                if (expanded) Icons.Outlined.KeyboardArrowUp else Icons.Outlined.KeyboardArrowDown,
                contentDescription = null, tint = Ink3
            )
        }

        if (expanded) {
            if (availableGrades.isEmpty()) {
                Text("Esta escuela todavía no tiene vías con grado.",
                    fontSize = 12.sp, color = Ink3,
                    modifier = Modifier.padding(horizontal = 16.dp))
            } else {
                LazyRow(
                    horizontalArrangement = Arrangement.spacedBy(6.dp),
                    modifier = Modifier.padding(horizontal = 16.dp)
                ) {
                    items(availableGrades) { g ->
                        GradeChip(
                            grade = g,
                            active = selectedGrades.contains(g),
                            onClick = {
                                onSelectedGradesChange(
                                    if (selectedGrades.contains(g)) selectedGrades - g else selectedGrades + g
                                )
                            }
                        )
                    }
                }

                if (isActive) {
                    Text(
                        "Mostrando ${result.matchingLines} vías de ${result.totalLines}",
                        fontSize = 12.sp, color = Ink2,
                        modifier = Modifier.padding(horizontal = 16.dp, vertical = 6.dp)
                    )
                    Column(Modifier.padding(horizontal = 16.dp)) {
                        result.groups.forEach { (grade, matches) ->
                            GradeGroup(
                                grade = grade,
                                matches = matches,
                                open = openGroups.contains(grade),
                                onToggle = {
                                    openGroups = if (openGroups.contains(grade)) openGroups - grade
                                    else openGroups + grade
                                },
                                onSelectLine = onSelectLine
                            )
                        }
                    }
                }
            }
        }
        Spacer(Modifier.height(if (expanded) 10.dp else 0.dp))
    }
}

@Composable
private fun GradeChip(grade: String, active: Boolean, onClick: () -> Unit) {
    val accent = colorForGrade(grade).let { if (it == Color.White) Ink else it }
    Row(
        Modifier
            .background(if (active) accent else Paper, RoundedCornerShape(2.dp))
            .border(1.dp, accent, RoundedCornerShape(2.dp))
            .clickable(onClick = onClick)
            .padding(horizontal = 10.dp, vertical = 5.dp)
    ) {
        Text(grade.lowercase(), fontSize = 12.sp, fontWeight = androidx.compose.ui.text.font.FontWeight.Bold,
            color = if (active) Color.White else accent)
    }
}

@Composable
private fun GradeGroup(
    grade: String,
    matches: List<GradeMatch>,
    open: Boolean,
    onToggle: () -> Unit,
    onSelectLine: (GradeMatch) -> Unit
) {
    val accent = colorForGrade(grade).let { if (it == Color.White) Ink else it }
    Column {
        Row(
            Modifier
                .fillMaxWidth()
                .clickable(onClick = onToggle)
                .padding(vertical = 8.dp),
            verticalAlignment = androidx.compose.ui.Alignment.CenterVertically
        ) {
            Row(Modifier
                .background(accent, RoundedCornerShape(2.dp))
                .padding(horizontal = 9.dp, vertical = 3.dp)) {
                Text(grade.lowercase(), fontSize = 12.sp,
                    fontWeight = androidx.compose.ui.text.font.FontWeight.Bold, color = Color.White)
            }
            Spacer(Modifier.width(8.dp))
            Text(if (matches.size == 1) "1 vía" else "${matches.size} vías", fontSize = 12.sp, color = Ink3)
            Spacer(Modifier.weight(1f))
            Icon(
                if (open) Icons.Outlined.KeyboardArrowDown else Icons.Outlined.KeyboardArrowRight,
                contentDescription = null, tint = Ink3
            )
        }
        if (open) {
            Column {
                matches.forEach { match ->
                    Row(
                        Modifier
                            .fillMaxWidth()
                            .clickable { onSelectLine(match) }
                            .padding(vertical = 8.dp),
                        verticalAlignment = androidx.compose.ui.Alignment.CenterVertically
                    ) {
                        Column(Modifier.weight(1f)) {
                            Text(match.lineName, fontSize = 13.sp,
                                fontWeight = androidx.compose.ui.text.font.FontWeight.SemiBold, color = Ink)
                            Text(match.blockName, fontSize = 11.sp, color = Ink3)
                        }
                        Icon(Icons.Outlined.KeyboardArrowRight, contentDescription = null, tint = Terra)
                    }
                    HorizontalDivider(color = Rule)
                }
            }
        }
    }
}

File: iosApp/iosApp/Screens/GradeFilterBar.swift
import SwiftUI
import Shared

// Filtro LOCAL por grado dentro de una escuela — ver BLOCK_SEARCH_DESIGN.md §7.
// Chips multi-selección con la paleta de grados de la app (mismo patrón que
// JournalBlocksListView), + lista de resultados agrupada por grado y
// colapsable. La lógica de qué piedra/vía cae en la selección vive en shared
// (GradeFilter.kt, `filterBlocksByGrades`) — este fichero es solo interfaz.

struct GradeFilterBar: View {
    @Binding var selectedGrades: Set<String>
    let availableGrades: [String]
    let result: GradeFilterResult
    let onSelectLine: (GradeMatch) -> Void

    @State private var expanded = false
    /// Grados con su grupo de resultados desplegado. Se conserva mientras la
    /// escuela siga abierta (aunque se navegue a una vía y se vuelva atrás).
    @State private var openGroups: Set<String> = []

    private var isActive: Bool { !selectedGrades.isEmpty }

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            Button {
                withAnimation(.easeInOut(duration: 0.15)) { expanded.toggle() }
            } label: {
                HStack(spacing: 6) {
                    Image(systemName: "slider.horizontal.3")
                        .font(.system(size: 13))
                        .foregroundStyle(isActive ? Cumbre.terra : Cumbre.ink3)
                    Text("FILTRAR POR GRADO")
                        .font(Cumbre.mono(10, .bold)).tracking(1.4)
                        .foregroundStyle(isActive ? Cumbre.terra : Cumbre.ink3)
                    Spacer()
                    if isActive {
                        Button {
                            selectedGrades = []
                        } label: {
                            Text("QUITAR TODO")
                                .font(Cumbre.mono(9, .bold)).tracking(1)
                                .foregroundStyle(Cumbre.ink3)
                        }.buttonStyle(.plain)
                    }
                    Image(systemName: expanded ? "chevron.up" : "chevron.down")
                        .font(.system(size: 12))
                        .foregroundStyle(Cumbre.ink3)
                }
                .padding(.horizontal, 16).padding(.vertical, 8)
                .contentShape(Rectangle())
            }
            .buttonStyle(.plain)

            if expanded {
                if availableGrades.isEmpty {
                    Text("Esta escuela todavía no tiene vías con grado.")
                        .font(.system(size: 12)).foregroundStyle(Cumbre.ink3)
                        .padding(.horizontal, 16)
                } else {
                    ScrollView(.horizontal, showsIndicators: false) {
                        HStack(spacing: 6) {
                            ForEach(availableGrades, id: \.self) { g in
                                gradeChip(g)
                            }
                        }
                        .padding(.horizontal, 16)
                    }

                    if isActive {
                        Text("Mostrando \(result.matchingLines) vías de \(result.totalLines)")
                            .font(.system(size: 12)).foregroundStyle(Cumbre.ink2)
                            .padding(.horizontal, 16)

                        resultsList
                    }
                }
            }
        }
        .padding(.bottom, expanded ? 10 : 0)
    }

    @ViewBuilder
    private func gradeChip(_ g: String) -> some View {
        let st = GradeColor.style(g)
        let accent = st.dark ? Cumbre.ink : st.stroke
        let active = selectedGrades.contains(g)
        Button {
            if active { selectedGrades.remove(g) } else { selectedGrades.insert(g) }
        } label: {
            Text(g.lowercased())
                .font(.system(size: 12, weight: .bold))
                .foregroundStyle(active ? .white : accent)
                .padding(.horizontal, 10).padding(.vertical, 5)
                .background(RoundedRectangle(cornerRadius: 2)
                    .fill(active ? accent : Cumbre.paper))
                .overlay(RoundedRectangle(cornerRadius: 2)
                    .stroke(accent, lineWidth: 1))
        }
        .buttonStyle(.plain)
    }

    @ViewBuilder
    private var resultsList: some View {
        VStack(alignment: .leading, spacing: 0) {
            ForEach(result.groups, id: \.0) { grade, matches in
                gradeGroup(grade: grade, matches: matches)
            }
        }
        .padding(.horizontal, 16)
        .padding(.top, 4)
    }

    @ViewBuilder
    private func gradeGroup(grade: String, matches: [GradeMatch]) -> some View {
        let st = GradeColor.style(grade)
        let accent = st.dark ? Cumbre.ink : st.stroke
        let isOpen = openGroups.contains(grade)

        Button {
            withAnimation(.easeInOut(duration: 0.12)) {
                if isOpen { openGroups.remove(grade) } else { openGroups.insert(grade) }
            }
        } label: {
            HStack(spacing: 8) {
                Text(grade.lowercased())
                    .font(.system(size: 12, weight: .bold))
                    .foregroundStyle(.white)
                    .padding(.horizontal, 9).padding(.vertical, 3)
                    .background(RoundedRectangle(cornerRadius: 2).fill(accent))
                Text(matches.count == 1 ? "1 vía" : "\(matches.count) vías")
                    .font(.system(size: 12)).foregroundStyle(Cumbre.ink3)
                Spacer()
                Image(systemName: isOpen ? "chevron.down" : "chevron.right")
                    .font(.system(size: 11)).foregroundStyle(Cumbre.ink3)
            }
            .padding(.vertical, 8)
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)

        if isOpen {
            VStack(spacing: 0) {
                ForEach(matches, id: \.lineId) { match in
                    Button {
                        onSelectLine(match)
                    } label: {
                        HStack {
                            VStack(alignment: .leading, spacing: 1) {
                                Text(match.lineName)
                                    .font(.system(size: 13, weight: .semibold))
                                    .foregroundStyle(Cumbre.ink)
                                Text(match.blockName)
                                    .font(.system(size: 11)).foregroundStyle(Cumbre.ink3)
                            }
                            Spacer()
                            Image(systemName: "chevron.right")
                                .font(.system(size: 11)).foregroundStyle(Cumbre.terra)
                        }
                        .padding(.vertical, 8)
                        .contentShape(Rectangle())
                    }
                    .buttonStyle(.plain)
                    Divider().background(Cumbre.rule)
                }
            }
        }
    }
}
